src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+
+/* input precision is how many fraction bits the input value has; this is a 64-bit number */
+#define RECIPLOG_INPUT_PREC		32
+
+/* lookup precision is how many fraction bits each table entry contains */
+#define RECIPLOG_LOOKUP_PREC	22
+
+/* output precision is how many fraction bits the result should have */
+#define RECIP_OUTPUT_PREC		15
+#define LOG_OUTPUT_PREC			8
+
+
+/*************************************
+ *
+ *  Dithering tables
+ *
+ *************************************/
+
+static const UINT8 dither_matrix_4x4[16] =
+{
+	 0,  8,  2, 10,
+	12,  4, 14,  6,
+	 3, 11,  1,  9,
+	15,  7, 13,  5
+};
+
+static const UINT8 dither_matrix_2x2[16] =
+{
+	 2, 10,  2, 10,
+	14,  6, 14,  6,
+	 2, 10,  2, 10,
+	14,  6, 14,  6
+};
+
+
+/*************************************
+ *
+ *  Macros for extracting pixels
+ *
+ *************************************/
+
+#define EXTRACT_565_TO_888(val, a, b, c)					\
+	(a) = (((val) >> 8) & 0xf8) | (((val) >> 13) & 0x07);	\
+	(b) = (((val) >> 3) & 0xfc) | (((val) >> 9) & 0x03);	\
+	(c) = (((val) << 3) & 0xf8) | (((val) >> 2) & 0x07);	\
+
+#define EXTRACT_x555_TO_888(val, a, b, c)					\
+	(a) = (((val) >> 7) & 0xf8) | (((val) >> 12) & 0x07);	\
+	(b) = (((val) >> 2) & 0xf8) | (((val) >> 7) & 0x07);	\
+	(c) = (((val) << 3) & 0xf8) | (((val) >> 2) & 0x07);	\
+
+#define EXTRACT_555x_TO_888(val, a, b, c)					\
+	(a) = (((val) >> 8) & 0xf8) | (((val) >> 13) & 0x07);	\
+	(b) = (((val) >> 3) & 0xf8) | (((val) >> 8) & 0x07);	\
+	(c) = (((val) << 2) & 0xf8) | (((val) >> 3) & 0x07);	\
+
+#define EXTRACT_1555_TO_8888(val, a, b, c, d)				\
+	(a) = ((INT16)(val) >> 15) & 0xff;						\
+	EXTRACT_x555_TO_888(val, b, c, d)						\
+
+#define EXTRACT_5551_TO_8888(val, a, b, c, d)				\
+	EXTRACT_555x_TO_888(val, a, b, c)						\
+	(d) = ((val) & 0x0001) ? 0xff : 0x00;					\
+
+#define EXTRACT_x888_TO_888(val, a, b, c)					\
+	(a) = ((val) >> 16) & 0xff;								\
+	(b) = ((val) >> 8) & 0xff;								\
+	(c) = ((val) >> 0) & 0xff;								\
+
+#define EXTRACT_888x_TO_888(val, a, b, c)					\
+	(a) = ((val) >> 24) & 0xff;								\
+	(b) = ((val) >> 16) & 0xff;								\
+	(c) = ((val) >> 8) & 0xff;								\
+
+#define EXTRACT_8888_TO_8888(val, a, b, c, d)				\
+	(a) = ((val) >> 24) & 0xff;								\
+	(b) = ((val) >> 16) & 0xff;								\
+	(c) = ((val) >> 8) & 0xff;								\
+	(d) = ((val) >> 0) & 0xff;								\
+
+#define EXTRACT_4444_TO_8888(val, a, b, c, d)				\
+	(a) = (((val) >> 8) & 0xf0) | (((val) >> 12) & 0x0f);	\
+	(b) = (((val) >> 4) & 0xf0) | (((val) >> 8) & 0x0f);	\
+	(c) = (((val) >> 0) & 0xf0) | (((val) >> 4) & 0x0f);	\
+	(d) = (((val) << 4) & 0xf0) | (((val) >> 0) & 0x0f);	\
+
+#define EXTRACT_332_TO_888(val, a, b, c)					\
+	(a) = (((val) >> 0) & 0xe0) | (((val) >> 3) & 0x1c) | (((val) >> 6) & 0x03); \
+	(b) = (((val) << 3) & 0xe0) | (((val) >> 0) & 0x1c) | (((val) >> 3) & 0x03); \
+	(c) = (((val) << 6) & 0xc0) | (((val) << 4) & 0x30) | (((val) << 2) & 0xc0) | (((val) << 0) & 0x03); \
+
```

Comment:
Define galore... 🥶

Reply:
Eh, yeah... Forcing the inlining this way might be beneficial though.

Agreed, to be fixed in future PR.